### ISSUE TITLE
Neural Net on Images 

### ISSUE THREAD

**Issue by Rohan2821999** [Rohan2821999/Neural-Nets#2]
@cbattista, I ran the images data on a super simplified botzmann NN model (not a deep layer network) unsupervised. The acc is around 60%. I hence recreated the images from the output data. 

Below is a comparison:

Actual Image Input:
![image](https://cloud.githubusercontent.com/assets/13100688/18604475/f8821980-7c31-11e6-8446-372c92fb66d1.png)

Reproduced NN Image..

![image](https://cloud.githubusercontent.com/assets/13100688/18604459/b59cc7f0-7c31-11e6-912e-40fcdaa8becf.png)

Looks to work pretty close for the first run. However, its not super accurate for some cases like outputting 7 etc. I am trying to figure why that is the case!

**Comment by Rohan2821999** [comment]
Since, this is a single layer network (single bias and weight) I am a bit unclear about how to figure out or show which neurons are active for specific numbers..

**Comment by cbattista** [comment]
Cool - I think it's time for me to dig into some code here.  Which file in
the repo is doing this - ImgNN.py, right?

**Comment by Rohan2821999** [comment]
Yup! It's that one.. I  am training the data directly on the xinput so all nodes are directly connected..

**Comment by Rohan2821999** [comment]
Figured out how Zorzi network works (Below are the layers of computation):
1. Layer1 Computation = Xinput matrix X Weight1 matrix + Bias matrix 
2. Layer2 Computation = Layer1 output X Weight2 matrix + Bias matrix --> activation values 

The above layers are trained according to contrastive divergence (which is basically a complicated name for averaged gradient descent) to give the desired activation values after 'n' epochs. 

The activation values from above neural net are fed into a neural net (backpropogated and reconstructed) 

Reconstruct1  = Activation matrix X Weight2 + Bias2 
Reconstruct2 = Reconstruct1 X Weight1 + Bias1 --> desired output close to x values based on neurons

A separate neural net is needed for the above reconstructions since a separate number of epochs of training is required.

I have therefore mimicked exactly the zorzi's neural net however I am unclear about dimensions of matrices he has used (just copied from there to initially get it working). Also, I am unsure on how to run two separate neural nets nested under a single network (would look into that). 

I think we could try to run it on Sherlock tomorrow (if possible)

**Comment by cbattista** [comment]
Great.  Jonathan will give us a tour of Sherlock at 1pm tomorrow!

**Comment by Rohan2821999** [comment]
@cbattista ... So, I built a generate Bayesian based neural net model where the weights and the hidden layers get updated from probability distributions of features generated from the Xinput (autonencoder RBMM based with 2 hidden layers).. It's a completely unsupervised model (unlike before which was kind of supervised) and works for unseen data as well.(works decently). 

The weights are randomly initialized from a gaussian curve every time the model is run so the output is not always the same however, I could definitely figure out which values of initial weights yields the best output.  

Below is the sample output of image produced:

Unsupervised model output:

![image](https://cloud.githubusercontent.com/assets/13100688/18768093/c6594730-80d7-11e6-864b-fd71f2ffe2d4.png)

Corresponding Xinput:
![image](https://cloud.githubusercontent.com/assets/13100688/18768058/a16072e6-80d7-11e6-8d43-c5b811873d87.png)

The spatial locations of not all circles match however, the number is more or less the same. 

I would send you the histograms of hidden layers for specific numbers later today..

Do you think something like above would be good?

**Comment by cbattista** [comment]
Yeah, that looks pretty good.

**Comment by Rohan2821999** [comment]
Okay! So, I ran the network by training individual set of images (all types of 3 dots only). And then I checked which neurons activated the most - the neuron number is not all always the same. However, there are a set of neurons which when activated the most would yield 3 dots output.

For example, sometimes neuron 25 gets activated, other times neuron 45 gets activated etc.  and when these specific neurons get activated it yields good results. However, though if neuron 22 is activated and backpropogated then the corresponding output image is bad..

**Comment by Rohan2821999** [comment]
In the above case, I am training only a set of number dot images (for example only 3 above) to get the histogram. I don't think that's a good way to do it - I think we should be training the entire dataset together but then I don't know how to get a histogram for specific numbers when the entire dataset is inputted.

**Comment by cbattista** [comment]
Yeah, you definitely need to train it on the entire dataset.

Then, once it's train, you just show it one image and get the activation in
the hidden units for that particular image.

-C

**Comment by cbattista** [comment]
*trained

**Comment by Rohan2821999** [comment]
I am not sure how to do that. I can train the entire data set but I am not sure how to just show it one image and get the activation.

**Comment by Rohan2821999** [comment]
The local minima issue is not resolved by the workaround we thought about.. Once it gets stuck changing the step is not making a difference and nor is randomly changing a given set of weights..

So, I hold some ambiguity whether it really is a local minima issue or something else!
